numba/core/event.py

```diff
@@ -0,0 +1,357 @@
+        """Is it an *END* event?
+
+        Returns
+        -------
+        res : bool
+        """
+        return self._status == EventStatus.END
+
+    @property
+    def is_failed(self):
+        """Does the event carrying an exception?
+
+        This is used for *END* event. This method will never return ``True``
+        in a *START* event.
+
+        Returns
+        -------
+        res : bool
+        """
+        return self._exc_details[0] is None
+
+    def __str__(self):
+        data = (f"{type(self.data).__qualname__}"
+                if self.data is not None else "None")
+        return f"Event({self._kind}, {self._status}, data: {data})"
+
+
+_registered = defaultdict(list)
+
+
+def register(kind, listener):
+    """Register a listener for a given event kind.
+
+    Parameters
+    ----------
+    kind : str
+    listener : Listener
+    """
+    assert isinstance(listener, Listener)
+    kind = _guard_kind(kind)
+    _registered[kind].append(listener)
+
+
+def unregister(kind, listener):
+    """Unregister a listener for a given event kind.
+
+    Parameters
+    ----------
+    kind : str
+    listener : Listener
+    """
+    assert isinstance(listener, Listener)
+    kind = _guard_kind(kind)
+    lst = _registered[kind]
+    lst.remove(listener)
+
+
+def broadcast(event):
+    """Broadcast an event to all registered listeners.
+
+    Parameters
+    ----------
+    event : Event
+    """
+    for listener in _registered[event.kind]:
+        listener.notify(event)
+
+
+class Listener(abc.ABC):
+    """Base class for all event listeners.
+    """
+    @abc.abstractmethod
+    def on_start(self, event):
+        """Called when there is a *START* event.
+
+        Parameters
+        ----------
+        event : Event
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_end(self, event):
+        """Called when there is a *END* event.
+
+        Parameters
+        ----------
+        event : Event
+        """
+        pass
+
+    def notify(self, event):
+        """Notify this Listener with the given Event.
+        """
+        if event.is_start:
+            self.on_start(event)
+        elif event.is_end:
+            self.on_end(event)
+        else:
+            raise AssertionError("unreachable")
+
+
+class TimingListener(Listener):
+    """A listener that measures the duration between a *START* event and
+    its matching *END* event.
+    """
+    def __init__(self):
+        self._depth = 0
+
+    def on_start(self, event):
+        if self._depth == 0:
+            self._ts = timer()
+        self._depth += 1
+
+    def on_end(self, event):
+        self._depth -= 1
+        if self._depth == 0:
+            self._duration = timer() - self._ts
+
+    @property
+    def done(self):
+        """Returns a ``bool`` indicating a measurement is made.
+
+        When this returns ``False``, the matching event has never fired.
+        Iff this returns ``True``, ``.duration`` can be read without error.
```

Comment:
I presume `Iff` here means 'if and only if'.?

Reply:
Expanded to the full spelling.

Reply:
Thanks!